File: docs/demo/spancomputer.js
// Port JS STRICT de adapter-vsto/src/MathCursor/Host/SpanComputer.cs
// Calcul pur de la span maths autour du caret (chemin Ctrl+Espace du produit) :
// bornes = delimiteur de phrase / stopword / OMath (ici : zone deja rendue) /
// debut-fin de ¶. Parite verrouillee par spancomputer.test.js (cas portes de
// SpanComputerTests.cs). Toute modif ici DOIT rester synchro avec le C#.
//
// NB : ce chemin ne sert qu'au declenchement explicite. L'auto-detection produit
// passe par le NER (non embarque en WASM) ; la demo s'en tient au caret-span.
(function (root) {
  'use strict';

  // Stopwords FR (table portee telle quelle du C#). Comparaison insensible a la
  // casse -> on stocke en minuscules et on lowercase le mot teste.
  var STOPWORDS = new Set([
    'soit', 'soient', 'et', 'ou', 'donc', 'alors', 'avec', 'si', 'on',
    'car', 'mais', 'ainsi', 'puis', 'comme', 'tout', 'un', 'une',
    'le', 'la', 'les', 'des', 'du', 'de', 'pour', 'par', 'sur',
    'dans', 'au', 'aux'
  ]);

  // '!' VOLONTAIREMENT ABSENT : postfixe factoriel, pas une fin de phrase
  // (ADR 2026-06-18-Fix-input-autocorrect-fraction-factorial). ',' n'est pas un
  // delimiteur (seul ';' coupe hors groupe) — le code garde le test ';'/',' du C#.
  var SPAN_DELIMITERS = new Set(['.', ';', '?', '=', '<', '>', '\n', '\r']);

  // Set REDUIT pour la demo « mode reel » : on retire les relations = < > pour
  // que l'equation entiere (ex. f(x)=1/x) soit captee d'un coup, faute de NER
  // pour la detecter. Les bornes de phrase/structure (. ; ? saut de ligne) et
  // les stopwords restent actifs. Cf. ADR 2026-06-19-Feat-web-demo-real-mode-editor.
  var DEMO_DELIMITERS = new Set(['.', ';', '?', '\n', '\r']);

  // char.IsLetter (unicode) || apostrophe || tiret.
  function isWordChar(c) {
    return /\p{L}/u.test(c) || c === "'" || c === '-';
  }

  function isWhitespace(c) {
    return /\s/.test(c);
  }

  // Position de l'ouvrante ( ou [ NON fermee qui englobe le caret (groupe en
  // cours de frappe), ou -1. Ne traverse pas un saut de ligne. Le '.' n'arrete
  // PAS le scan (separateur decimal). Cf. EnclosingOpenBracket (C#).
  function enclosingOpenBracket(text, caret) {
    var depth = 0;
    for (var k = caret - 1; k >= 0; k--) {
      var c = text[k];
      if (c === '\n' || c === '\r') return -1;
      if (c === ')' || c === ']') { depth++; continue; }
      if (c === '(' || c === '[') {
        if (depth > 0) { depth--; continue; }
        return k;
      }
    }
    return -1;
  }

  // Nombre d'ouvrantes ( / [ NON fermees avant le caret (remis a zero a chaque
  // saut de ligne). Init de la marche avant de computeSpanEnd. Cf. OpenDepthBehind.
  function openDepthBehind(text, caret) {
    var parenOpen = 0, bracketOpen = 0;
    var n = Math.min(caret, text.length);
    for (var k = 0; k < n; k++) {
      var c = text[k];
      if (c === '\n' || c === '\r') { parenOpen = 0; bracketOpen = 0; continue; }
      if (c === '(') parenOpen++;
      else if (c === ')') { if (parenOpen > 0) parenOpen--; }
      else if (c === '[') bracketOpen++;
      else if (c === ']') { if (bracketOpen > 0) bracketOpen--; }
    }
    return { parenOpen: parenOpen, bracketOpen: bracketOpen };
  }

  // omathRegions : tableau de { start, end } (zones deja rendues = bornes dures,
  // role des OMaths Word). Peut etre null/[].
  // delims : Set de delimiteurs (defaut = SPAN_DELIMITERS, = parite C#). La demo
  // « mode reel » passe un set REDUIT (sans = < >) : sans NER pour detecter
  // l'equation entiere, on ne veut pas que les relations coupent la zone.
  function computeSpanStart(text, caret, omathRegions, delims) {
    delims = delims || SPAN_DELIMITERS;
    var start = 0;

    var openBracket = enclosingOpenBracket(text, caret);
    if (openBracket >= 0) {
      start = openBracket;
    } else {
      var bracketDepth = 0, parenDepth = 0;
      for (var k = caret - 1; k >= 0; k--) {
        var c = text[k];
        if (c === ']') { bracketDepth++; continue; }
        if (c === '[') { if (bracketDepth > 0) bracketDepth--; continue; }
        if (c === ')') { parenDepth++; continue; }
        if (c === '(') { if (parenDepth > 0) parenDepth--; continue; }

        if (!delims.has(c)) continue;
        if ((c === ';' || c === ',') && (bracketDepth > 0 || parenDepth > 0)) continue;
        start = Math.max(start, k + 1);
        break;
      }
    }

    if (omathRegions) {
      for (var r = 0; r < omathRegions.length; r++) {
        if (omathRegions[r].end <= caret) start = Math.max(start, omathRegions[r].end);
      }
    }

    var i = caret - 1;
    while (i >= start) {
      while (i >= start && isWhitespace(text[i])) i--;
      if (i < start) break;
      var wordEnd = i + 1;
      while (i >= start && isWordChar(text[i])) i--;
      var wordStart = i + 1;
      if (wordEnd <= wordStart) { i--; continue; }
      var w = text.substring(wordStart, wordEnd);
      if (STOPWORDS.has(w.toLowerCase())) { start = wordEnd; break; }
    }

    return start;
  }

  function computeSpanEnd(text, caret, omathRegions, delims) {
    delims = delims || SPAN_DELIMITERS;
    var end = text.length;
    while (end > caret && (text[end - 1] === '\r' || text[end - 1] === '\n')) end--;

    var depth = openDepthBehind(text, caret);
    var parenDepth = depth.parenOpen, bracketDepth = depth.bracketOpen;
    for (var k = caret; k < end; k++) {
      var c = text[k];
      if (c === '[') { bracketDepth++; continue; }
      if (c === ']') { if (bracketDepth > 0) bracketDepth--; continue; }
      if (c === '(') { parenDepth++; continue; }
      if (c === ')') { if (parenDepth > 0) parenDepth--; continue; }

      if (!delims.has(c)) continue;
      if ((c === ';' || c === ',') && (bracketDepth > 0 || parenDepth > 0)) continue;
      end = k;
      break;
    }

    if (omathRegions) {
      for (var r = 0; r < omathRegions.length; r++) {
        var s = omathRegions[r].start;
        if (s >= caret && s < end) end = s;
      }
    }

    var i = caret;
    while (i < end) {
      while (i < end && isWhitespace(text[i])) i++;
      if (i >= end) break;
      var wordStart = i;
      while (i < end && isWordChar(text[i])) i++;
      var wordEnd = i;
      if (wordEnd <= wordStart) { i++; continue; }
      var w = text.substring(wordStart, wordEnd);
      if (STOPWORDS.has(w.toLowerCase())) { end = wordStart; break; }
    }

    return end;
  }

  // Reproduit le flux de ConversionController.Trigger : span brute puis trim des
  // blancs aux deux bouts. Retourne { start, end } (offsets dans text).
  function computeZone(text, caret, omathRegions, delims) {
    var s = computeSpanStart(text, caret, omathRegions, delims);
    var e = computeSpanEnd(text, caret, omathRegions, delims);
    while (s < e && isWhitespace(text[s])) s++;
    while (e > s && isWhitespace(text[e - 1])) e--;
    return { start: s, end: e };
  }

  var api = {
    computeSpanStart: computeSpanStart,
    computeSpanEnd: computeSpanEnd,
    computeZone: computeZone,
    enclosingOpenBracket: enclosingOpenBracket,
    openDepthBehind: openDepthBehind,
    isWordChar: isWordChar,
    STOPWORDS: STOPWORDS,
    SPAN_DELIMITERS: SPAN_DELIMITERS,
    DEMO_DELIMITERS: DEMO_DELIMITERS
  };

  if (typeof module !== 'undefined' && module.exports) module.exports = api;
  if (root) root.SpanComputer = api;
})(typeof window !== 'undefined' ? window : null);

File: adapter-vsto/src/MathCursor/Host/SpanComputer.cs
// MathCursor: capturing mathematical intent from linear keyboard input.
// Copyright (C) 2026  Côme de Percin
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System;
using System.Collections.Generic;

namespace MathCursor.Host
{
    /// <summary>
    /// Calcul PUR (sans dépendance Word) de la span autour du caret pour le
    /// trigger manuel Ctrl+Espace : bornes = délimiteur de phrase / stopword /
    /// OMath / début-fin de ¶. Extrait de <see cref="ConversionController"/>
    /// pour être testable en unitaire (le contrôleur, lui, porte l'interop Word).
    /// <para>
    /// NB : ce chemin sert UNIQUEMENT au Ctrl+Espace explicite. L'auto-détection
    /// passe par le NER (zones), pas par ces délimiteurs. Cf. ADR
    /// 2026-06-18-Fix-input-autocorrect-fraction-factorial : <c>!</c> a été
    /// RETIRÉ du set (opérateur postfixe factoriel, pas une ponctuation — sinon
    /// Ctrl+Espace après <c>n!</c> donnait une span vide → pas de popup).
    /// </para>
    /// </summary>
    internal static class SpanComputer
    {
        // ── Données de span : stopwords + délimiteurs FR (table portée de
        //    l'ex data/locale/fr.yml, le YAML appartenait à l'ancien moteur).
        private static readonly HashSet<string> Stopwords = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            "soit", "soient", "et", "ou", "donc", "alors", "avec", "si", "on",
            "car", "mais", "ainsi", "puis", "comme", "tout", "un", "une",
            "le", "la", "les", "des", "du", "de", "pour", "par", "sur",
            "dans", "au", "aux",
        };

        // '!' VOLONTAIREMENT ABSENT : c'est un postfixe (factorielle), pas une
        // fin de phrase. Cf. ADR 2026-06-18-Fix-input-autocorrect-fraction-factorial.
        private static readonly HashSet<char> SpanDelimiters = new HashSet<char>
        {
            '.', ';', '?', '=', '<', '>', '\n', '\r',
        };

        /// <summary>
        /// Début de la span : max(début ¶, après le dernier délimiteur hors
        /// brackets/parens AVANT le caret, fin du dernier OMath avant le caret,
        /// dernier stopword mot-entier). Logique reprise de
        /// l'ex-ManualTriggerController (comportement validé).
        /// </summary>
        public static int ComputeSpanStart(string text, int caret,
            IReadOnlyList<(int start, int end)> omathRegions)
        {
            int start = 0;

            // Groupe ( … ou [ … NON FERMÉ englobant le caret (matrice / tuple /
            // intervalle en cours de frappe) : la zone démarre à l'ouvrante, les
            // ; et , internes sont structurels (pas des fins de phrase). Cf. ADR
            // 2026-06-19-Fix-spancomputer-unclosed-bracket-matrix.
            int openBracket = EnclosingOpenBracket(text, caret);
            if (openBracket >= 0)
            {
                start = openBracket;
            }
            else
            {
                // Après le dernier délimiteur — walk backward, suivi profondeur.
                int bracketDepth = 0, parenDepth = 0;
                for (int k = caret - 1; k >= 0; k--)
                {
                    char c = text[k];
                    if (c == ']') { bracketDepth++; continue; }
                    if (c == '[') { if (bracketDepth > 0) bracketDepth--; continue; }
                    if (c == ')') { parenDepth++; continue; }
                    if (c == '(') { if (parenDepth > 0) parenDepth--; continue; }

                    if (!SpanDelimiters.Contains(c)) continue;
                    if ((c == ';' || c == ',') && (bracketDepth > 0 || parenDepth > 0)) continue;
                    start = Math.Max(start, k + 1);
                    break;
                }
            }

            // Après la fin du dernier OMath qui se termine avant le caret.
            if (omathRegions != null)
                foreach (var (s, e) in omathRegions)
                    if (e <= caret) start = Math.Max(start, e);

            // Après le dernier stopword (mot entier).
            int i = caret - 1;
            while (i >= start)
            {
                while (i >= start && char.IsWhiteSpace(text[i])) i--;
                if (i < start) break;
                int wordEnd = i + 1;
                while (i >= start && IsWordChar(text[i])) i--;
                int wordStart = i + 1;
                if (wordEnd <= wordStart) { i--; continue; }
                string w = text.Substring(wordStart, wordEnd - wordStart);
                if (Stopwords.Contains(w)) { start = wordEnd; break; }
            }

            return start;
        }

        /// <summary>
        /// Fin de la span : min(fin ¶ [sans \r virtuel], premier délimiteur
        /// hors brackets/parens APRÈS le caret, début du premier OMath après
        /// le caret, premier stopword mot-entier après le caret). Miroir
        /// avant de <see cref="ComputeSpanStart"/> — un caret replacé au
        /// milieu d'une expression la capture en entier.
        /// </summary>
        public static int ComputeSpanEnd(string text, int caret,
            IReadOnlyList<(int start, int end)> omathRegions)
        {
            int end = text.Length;
            while (end > caret && (text[end - 1] == '\r' || text[end - 1] == '\n')) end--;

            // Premier délimiteur — walk forward avec suivi profondeur brackets/parens.
            // Init = ouvrantes NON fermées avant le caret (caret replacé au milieu
            // d'un groupe ( … / [ … en cours de frappe) → les ; , du groupe
            // englobant ne coupent pas. Cf. ADR 2026-06-19-Fix-spancomputer-
            // unclosed-bracket-matrix.
            OpenDepthBehind(text, caret, out int parenDepth, out int bracketDepth);
            for (int k = caret; k < end; k++)
            {
                char c = text[k];
                if (c == '[') { bracketDepth++; continue; }
                if (c == ']') { if (bracketDepth > 0) bracketDepth--; continue; }
                if (c == '(') { parenDepth++; continue; }
                if (c == ')') { if (parenDepth > 0) parenDepth--; continue; }

                if (!SpanDelimiters.Contains(c)) continue;
                if ((c == ';' || c == ',') && (bracketDepth > 0 || parenDepth > 0)) continue;
                end = k;
                break;
            }

            // Début du premier OMath qui commence après le caret.
            if (omathRegions != null)
                foreach (var (s, _) in omathRegions)
                    if (s >= caret && s < end) end = s;

            // Premier stopword (mot entier) après le caret.
            int i = caret;
            while (i < end)
            {
                while (i < end && char.IsWhiteSpace(text[i])) i++;
                if (i >= end) break;
                int wordStart = i;
                while (i < end && IsWordChar(text[i])) i++;
                int wordEnd = i;
                if (wordEnd <= wordStart) { i++; continue; }
                string w = text.Substring(wordStart, wordEnd - wordStart);
                if (Stopwords.Contains(w)) { end = wordStart; break; }
            }

            return end;
        }

        /// <summary>
        /// Position de l'ouvrante <c>(</c> ou <c>[</c> NON fermée qui englobe le
        /// caret (groupe en cours de frappe), ou -1. On ne traverse pas un saut
        /// de ligne (un groupe ne s'étend pas sur plusieurs lignes). Le <c>.</c>
        /// n'arrête PAS le scan (sinon un séparateur décimal <c>(1,5 ;2,5</c>
        /// casserait la détection).
        /// </summary>
        private static int EnclosingOpenBracket(string text, int caret)
        {
            int depth = 0;
            for (int k = caret - 1; k >= 0; k--)
            {
                char c = text[k];
                if (c == '\n' || c == '\r') return -1;
                if (c == ')' || c == ']') { depth++; continue; }
                if (c == '(' || c == '[')
                {
                    if (depth > 0) { depth--; continue; }
                    return k; // ouvrante non appariée englobant le caret
                }
            }
            return -1;
        }

        /// <summary>
        /// Nombre d'ouvrantes <c>(</c> / <c>[</c> NON fermées avant le caret
        /// (remis à zéro à chaque saut de ligne). Sert à initialiser la marche
        /// avant de <see cref="ComputeSpanEnd"/> quand le caret est au milieu
        /// d'un groupe ouvert.
        /// </summary>
        private static void OpenDepthBehind(string text, int caret, out int parenOpen, out int bracketOpen)
        {
            parenOpen = 0;
            bracketOpen = 0;
            int n = Math.Min(caret, text.Length);
            for (int k = 0; k < n; k++)
            {
                char c = text[k];
                if (c == '\n' || c == '\r') { parenOpen = 0; bracketOpen = 0; continue; }
                if (c == '(') parenOpen++;
                else if (c == ')') { if (parenOpen > 0) parenOpen--; }
                else if (c == '[') bracketOpen++;
                else if (c == ']') { if (bracketOpen > 0) bracketOpen--; }
            }
        }

        private static bool IsWordChar(char c) => char.IsLetter(c) || c == '\'' || c == '-';
    }
}
